pkg/controller/compute/v1beta1/cluster.go

```diff
@@ -0,0 +1,287 @@
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.Secret.Namespace, Name: p.Spec.Secret.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+
+	gke, err := c.newServiceFn(ctx,
+		option.WithCredentialsJSON(s.Data[p.Spec.Secret.Key]),
+		option.WithScopes(container.CloudPlatformScope))
+	return &clusterExternal{cluster: *gke, projectID: p.Spec.ProjectID, kube: c.kube}, errors.Wrap(err, errNewClient)
+}
+
+type clusterExternal struct {
+	kube      client.Client
+	cluster   container.Service
+	projectID string
+}
+
+func (e *clusterExternal) Observe(ctx context.Context, mg resource.Managed) (resource.ExternalObservation, error) { // nolint:gocyclo
+	cr, ok := mg.(*v1beta1.GKECluster)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errNotCluster)
+	}
+
+	existing, err := e.cluster.Projects.Locations.Clusters.Get(meta.GetExternalName(mg)).Context(ctx).Do()
+	if err != nil {
+		return resource.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errGetCluster)
+	}
+
+	cr.Status.AtProvider = gke.GenerateObservation(*existing)
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	gke.LateInitializeSpec(&cr.Spec.ForProvider, *existing)
+	if !reflect.DeepEqual(currentSpec, &cr.Spec.ForProvider) {
+		if err := e.kube.Update(ctx, cr); err != nil {
+			return resource.ExternalObservation{}, errors.Wrap(err, errManagedUpdateFailed)
+		}
+	}
+
+	// If boostrap node pool is present then we want to delete it.
+	for _, pool := range existing.NodePools {
+		if pool != nil {
+			if pool.Name == gke.BootstrapNodePoolName {
+				boostrapNP := strings.Join([]string{meta.GetExternalName(mg), gke.BootstrapNodePoolName}, "/")
+				_, err = e.cluster.Projects.Locations.Clusters.NodePools.Delete(boostrapNP).Context(ctx).Do()
+				if err != nil {
+					// If we fail to delete bootstrap node pool we will be requeued implicitly.
+					return resource.ExternalObservation{}, errors.Wrap(resource.Ignore(gcp.IsErrorNotFound, err), errDeleteBootstrapNodePool)
+				}
+				break
+			}
+		}
+	}
+
+	switch cr.Status.AtProvider.Status {
+	case v1beta1.ClusterStateRunning:
+		cr.Status.SetConditions(v1alpha1.Available())
+		resource.SetBindable(cr)
+	case v1beta1.ClusterStateProvisioning:
+		cr.Status.SetConditions(v1alpha1.Creating())
+	case v1beta1.ClusterStateUnspecified, v1beta1.ClusterStateDegraded, v1beta1.ClusterStateError, v1beta1.ClusterStateReconciling:
+		cr.Status.SetConditions(v1alpha1.Unavailable())
+	}
+
+	upToDate, _ := gke.IsUpToDate(&cr.Spec.ForProvider, *existing)
+
+	return resource.ExternalObservation{
+		ResourceExists:    true,
+		ResourceUpToDate:  upToDate,
+		ConnectionDetails: connectionDetails(existing),
+	}, nil
+}
+
+func (e *clusterExternal) Create(ctx context.Context, mg resource.Managed) (resource.ExternalCreation, error) {
+	i, ok := mg.(*v1beta1.GKECluster)
```

Comment:
I think @muvaf has been slowly renaming all these variables to `cr` for "custom resource". We might want to be consistent here.

(I personally used try to use a short version of the thing's name, which would be `c` in this case since it's a "GKE cluster" and not a "GKE instance", but I'm happy to settle on `cr`.)